Improving distribution tests in conformance

On https://github.com/kubernetes-sigs/gateway-api/pull/3827#discussion_r2123121430

@guicassolato suggested the below: 

> Worth adding some entropy to each iteration to mitigate the risk of coordinated execution perhaps? For example: introducing jitter, randomising the payload, or anything else that varies the sample across iterations (without affecting expected weight distribution).

It would be nice to have something like thig

/good-first-issue

hi @LiorLieberman, my apologies for not submitting a PR yet, but I was spending the past few days trying to understand the tests and also running the test locally.

So far, I've ran the tests using the below command, without making any changes ( I wanted to make sure I was running the tests correctly and not break anything ).
```
go test -v ./conformance \
--run TestConformance/MeshHTTPRouteWeight \
--supported-features=Mesh,HTTPRoute
```

And the test is returning me failure even without adding any entropy: 

```
    httproute-weight.go:64: Traffic distribution test failed (10/10): backend "echo-v1" weighted traffic of 0.524 not within tolerance 0.7 (+/-0.050000)
        backend "echo-v2" weighted traffic of 0.476 not within tolerance 0.3 (+/-0.050000)
    httproute-weight.go:69: Weighted distribution tests failed
=== NAME  TestConformance/MeshHTTPRouteWeight
    apply.go:283: 2025-06-10T18:05:28.312926+08:00: Deleting mesh-weighted-backends HTTPRoute
=== NAME  TestConformance
    apply.go:283: 2025-06-10T18:05:28.317196+08:00: Deleting echo-v1 Deployment
    apply.go:283: 2025-06-10T18:05:28.320049+08:00: Deleting gateway-conformance-mesh-consumer Namespace
    apply.go:283: 2025-06-10T18:05:28.322049+08:00: Deleting echo Service
    apply.go:283: 2025-06-10T18:05:28.330848+08:00: Deleting echo-v2 Service
    apply.go:283: 2025-06-10T18:05:28.341497+08:00: Deleting echo-v2 Deployment
    apply.go:283: 2025-06-10T18:05:28.345876+08:00: Deleting echo-v1 Service
    apply.go:283: 2025-06-10T18:05:28.352877+08:00: Deleting echo-v1 Deployment
    apply.go:283: 2025-06-10T18:05:28.357337+08:00: Deleting gateway-conformance-mesh Namespace
--- FAIL: TestConformance (73.35s)
    --- FAIL: TestConformance/MeshHTTPRouteWeight (71.08s)
        --- FAIL: TestConformance/MeshHTTPRouteWeight/Requests_should_have_a_distribution_that_matches_the_weight (71.06s)
```

I'm not sure if I'm running the tests wrongly, but could you point me to the right direction? thanks !

--- a/conformance/tests/mesh/httproute-weight.go
+++ b/conformance/tests/mesh/httproute-weight.go
@@ -21,10 +21,12 @@ import (
 	"errors"
 	"fmt"
 	"math"
+	"math/rand"
 	"slices"
 	"strings"
 	"sync"
 	"testing"
+	"time"
 
 	"golang.org/x/sync/errgroup"
 
@@ -52,7 +54,7 @@ var MeshHTTPRouteWeight = suite.ConformanceTest{
 		t.Run("Requests should have a distribution that matches the weight", func(t *testing.T) {
 			host := "echo"
 			expected := http.ExpectedResponse{
-				Request:   http.Request{Path: "/", Host: host},
+				Request:   http.Request{Path: "/", Host: host, Headers: make(map[string]string)},
 				Response:  http.Response{StatusCode: 200},
 				Namespace: "gateway-conformance-mesh",
 			}
@@ -89,7 +91,9 @@ func testDistribution(t *testing.T, client echo.MeshPod, expected http.ExpectedR
 	g.SetLimit(concurrentRequests)
 	for i := 0.0; i < totalRequests; i++ {
 		g.Go(func() error {
-			_, cRes, err := client.CaptureRequestResponseAndCompare(t, expected)
+			uniqueExpected := expected
+			addEntropy(&uniqueExpected)
+			_, cRes, err := client.CaptureRequestResponseAndCompare(t, uniqueExpected)
 			if err != nil {
 				return fmt.Errorf("failed: %w", err)
 			}
@@ -140,4 +144,19 @@ func testDistribution(t *testing.T, client echo.MeshPod, expected http.ExpectedR
 		return cmp.Compare(a.Error(), b.Error())
 	})
 	return errors.Join(errs...)
+}
+
+// addEntropy adds jitter to the request by adding either a delay up to 1 second, or a random header value, or both.
+func addEntropy(exp *http.ExpectedResponse) {
+	delay := func() { time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond) }
+	randomHeader := func() { exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", rand.Intn(9999)) }
+	switch rand.Intn(3) {
+	case 0:
+		delay()
+	case 1:
+		randomHeader()
+	case 2:
+		delay()
+		randomHeader()
+	}
 }

--- a/conformance/tests/mesh/httproute-weight.go
+++ b/conformance/tests/mesh/httproute-weight.go
@@ -148,9 +148,9 @@ func testDistribution(t *testing.T, client echo.MeshPod, expected http.ExpectedR
 
 // addEntropy adds jitter to the request by adding either a delay up to 1 second, or a random header value, or both.
 func addEntropy(exp *http.ExpectedResponse) {
-	delay := func() { time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond) }
-	randomHeader := func() { exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", rand.Intn(9999)) }
-	switch rand.Intn(3) {
+	delay := func() { time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond) }               //nolint:gosec // This is test code to get random value.
+	randomHeader := func() { exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", rand.Intn(9999)) } //nolint:gosec // This is test code to get random value.
+	switch rand.Intn(3) {                                                                           //nolint:gosec // This is test code to get random value
 	case 0:
 		delay()
 	case 1:

--- a/conformance/tests/mesh/httproute-weight.go
+++ b/conformance/tests/mesh/httproute-weight.go
@@ -196,6 +196,6 @@ func addEntropy(exp *http.ExpectedResponse) error {
 		}
 		return randomHeader(10000)
 	default:
-		return nil
+		return fmt.Errorf("invalid random value: %d", *random)
 	}
 }

--- a/conformance/tests/mesh/httproute-weight.go
+++ b/conformance/tests/mesh/httproute-weight.go
@@ -186,13 +186,16 @@ func addEntropy(exp *http.ExpectedResponse) error {
 	}
 
 	switch *random {
-	case int64(0):
-		delay(1000)
+	case 0:
+		return delay(1000)
 	case 1:
-		randomHeader(10000)
+		return randomHeader(10000)
 	case 2:
-		delay(1000)
-		randomHeader(10000)
+		if err := delay(1000); err != nil {
+			return err
+		}
+		return randomHeader(10000)
+	default:
+		return nil
 	}
-	return nil
 }

--- a/conformance/tests/mesh/httproute-weight.go
+++ b/conformance/tests/mesh/httproute-weight.go
@@ -18,10 +18,11 @@ package meshtests
 
 import (
 	"cmp"
+	"crypto/rand"
 	"errors"
 	"fmt"
 	"math"
-	"math/rand"
+	"math/big"
 	"slices"
 	"strings"
 	"sync"
@@ -92,7 +93,9 @@ func testDistribution(t *testing.T, client echo.MeshPod, expected http.ExpectedR
 	for i := 0.0; i < totalRequests; i++ {
 		g.Go(func() error {
 			uniqueExpected := expected
-			addEntropy(&uniqueExpected)
+			if err := addEntropy(&uniqueExpected); err != nil {
+				return fmt.Errorf("error adding entropy: %w", err)
+			}
 			_, cRes, err := client.CaptureRequestResponseAndCompare(t, uniqueExpected)
 			if err != nil {
 				return fmt.Errorf("failed: %w", err)
@@ -147,19 +150,49 @@ func testDistribution(t *testing.T, client echo.MeshPod, expected http.ExpectedR
 }
 
 // addEntropy adds jitter to the request by adding either a delay up to 1 second, or a random header value, or both.
-func addEntropy(exp *http.ExpectedResponse) {
-	delay := func() { time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond) } //nolint:gosec // This is test code to get random value.
-	randomHeader := func() {
+func addEntropy(exp *http.ExpectedResponse) error {
+	randomNumber := func(limit int64) (*int64, error) {
+		number, err := rand.Int(rand.Reader, big.NewInt(limit))
+		if err != nil {
+			return nil, err
+		}
+		n := number.Int64()
+		return &n, nil
+	}
+
+	// adds a delay
+	delay := func(limit int64) error {
+		randomSleepDuration, err := randomNumber(limit)
+		if err != nil {
+			return err
+		}
+		time.Sleep(time.Duration(*randomSleepDuration) * time.Millisecond)
+		return nil
+	}
+	// adds random header value
+	randomHeader := func(limit int64) error {
+		randomHeaderValue, err := randomNumber(limit)
+		if err != nil {
+			return err
+		}
 		exp.Request.Headers = make(map[string]string)
-		exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", rand.Intn(9999)) //nolint:gosec // This is test code to get random value.
+		exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", *randomHeaderValue)
+		return nil
 	}
-	switch rand.Intn(3) { //nolint:gosec // This is test code to get random value
-	case 0:
-		delay()
+
+	random, err := randomNumber(3)
+	if err != nil {
+		return err
+	}
+
+	switch *random {
+	case int64(0):
+		delay(1000)
 	case 1:
-		randomHeader()
+		randomHeader(10000)
 	case 2:
-		delay()
-		randomHeader()
+		delay(1000)
+		randomHeader(10000)
 	}
+	return nil
 }

--- a/conformance/tests/mesh/httproute-weight.go
+++ b/conformance/tests/mesh/httproute-weight.go
@@ -54,7 +54,7 @@ var MeshHTTPRouteWeight = suite.ConformanceTest{
 		t.Run("Requests should have a distribution that matches the weight", func(t *testing.T) {
 			host := "echo"
 			expected := http.ExpectedResponse{
-				Request:   http.Request{Path: "/", Host: host, Headers: make(map[string]string)},
+				Request:   http.Request{Path: "/", Host: host},
 				Response:  http.Response{StatusCode: 200},
 				Namespace: "gateway-conformance-mesh",
 			}
@@ -148,9 +148,12 @@ func testDistribution(t *testing.T, client echo.MeshPod, expected http.ExpectedR
 
 // addEntropy adds jitter to the request by adding either a delay up to 1 second, or a random header value, or both.
 func addEntropy(exp *http.ExpectedResponse) {
-	delay := func() { time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond) }               //nolint:gosec // This is test code to get random value.
-	randomHeader := func() { exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", rand.Intn(9999)) } //nolint:gosec // This is test code to get random value.
-	switch rand.Intn(3) {                                                                           //nolint:gosec // This is test code to get random value
+	delay := func() { time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond) } //nolint:gosec // This is test code to get random value.
+	randomHeader := func() {
+		exp.Request.Headers = make(map[string]string)
+		exp.Request.Headers["X-Jitter"] = fmt.Sprintf("%d", rand.Intn(9999)) //nolint:gosec // This is test code to get random value.
+	}
+	switch rand.Intn(3) { //nolint:gosec // This is test code to get random value
 	case 0:
 		delay()
 	case 1:

